ocaml/xapi/xapi_psr.ml

```diff
@@ -0,0 +1,143 @@
+type r =
+  | Success
+  | Failed_during_accept_new_pool_secret
+  | Failed_during_send_new_pool_secret
+  | Failed_during_finish
+
+module type Impl = sig
+  type pool_secret
+
+  type pool_secrets = pool_secret * pool_secret (* (old, new) *)
+
+  type host
+
+  val save_checkpoint : string -> unit
+
+  val retrieve_checkpoint : unit -> string option
+
+  val backup : pool_secrets -> unit
+
+  val retrieve : unit -> pool_secrets
+
+  val tell_accept_new_pool_secret : pool_secrets -> host -> unit
+
+  val tell_send_new_pool_secret : pool_secrets -> host -> unit
+
+  val tell_cleanup_old_pool_secret : host -> unit
+end
+
+module Make =
+functor
+  (Impl : Impl)
+  ->
+  struct
+    open Impl
+
+    type checkpoint =
+      | No_checkpoint
+      | Backup
+      | Accept_new_pool_secret
+      | Send_new_pool_secret
+      | Finish
+
+    exception Cannot_parse_checkpoint of string
+
+    let string_of_checkpoint = function
+      | No_checkpoint ->
+          "No_checkpoint"
+      | Backup ->
+          "Backup"
+      | Accept_new_pool_secret ->
+          "Accept_new_pool_secret"
+      | Send_new_pool_secret ->
+          "Send_new_pool_secret"
+      | Finish ->
+          "Finish"
+
+    let checkpoint_of_string = function
+      | "No_checkpoint" ->
+          No_checkpoint
+      | "Backup" ->
+          Backup
+      | "Accept_new_pool_secret" ->
+          Accept_new_pool_secret
+      | "Send_new_pool_secret" ->
+          Send_new_pool_secret
+      | "Finish" ->
+          Finish
+      | s ->
+          raise (Cannot_parse_checkpoint s)
+
+    let rec go pool_secrets hosts = function
+      | No_checkpoint ->
+          backup pool_secrets ;
+          (go [@tailcall]) pool_secrets hosts Backup
+      | Backup -> (
+          let res =
+            try
+              List.iter (tell_accept_new_pool_secret pool_secrets) hosts ;
+              None
+            with e -> Some Failed_during_accept_new_pool_secret
+          in
+          match res with
+          | None ->
+              (go [@tailcall]) pool_secrets hosts Accept_new_pool_secret
+          | Some e ->
+              save_checkpoint (string_of_checkpoint Backup) ;
+              e
+        )
+      | Accept_new_pool_secret -> (
+          let res =
+            try
+              List.iter (tell_accept_new_pool_secret pool_secrets) hosts ;
+              None
+            with e -> Some Failed_during_accept_new_pool_secret
+          in
+          match res with
+          | None ->
+              (go [@tailcall]) pool_secrets hosts Send_new_pool_secret
+          | Some e ->
+              save_checkpoint (string_of_checkpoint Accept_new_pool_secret) ;
```

Comment:
I think the checkpoint should be saved when you enter the go function, prior to doing the matching, that way you don't need all these match None/Some on each state, and also you would save a checkpoint even if xapi or the host crashes in the middle of running the operation.

As for implementing the save checkpoint: perhaps the redolog code can be used here? That already has necessary logic for recovering from failures, etc. We could require the redolog to be enabled (which doens't necessarily mean that HA has to be enabled, in fact HA should be disabled...) before doing a PSR